src/uuid/uuidbuilder.rs

```diff
@@ -33,174 +32,174 @@ pub struct UUIDBuilder {
-    /// if the system time is either
-    /// * set to a point in time before UNIX Epoch, or
-    /// * set to a point in time later than UNIX Epoch + 0xFFFFFFFFFFFF seconds
-    ///
-    /// # Examples
-    ///
-    /// ```rust
-    /// use up_rust::UUIDBuilder;
-    ///
-    /// let uuid = UUIDBuilder::new().build();
-    /// assert!(uuid.is_uprotocol_uuid());
-    /// assert!(uuid.get_time().is_some());
-    /// ```
-    pub fn build(&self) -> UUID {
-        let now = SystemTime::now()
-            .duration_since(UNIX_EPOCH)
-            .expect("current system time is set to a point in time before UNIX Epoch");
-        let now_millis = u64::try_from(now.as_millis())
-            .expect("current system time is set to a point in time too far in the future");
-        self.build_with_instant(now_millis)
-    }
-
     /// Creates a new UUID for a given timestamp.
     ///
-    /// # Arguments
+    /// # Note
     ///
-    /// * `timestamp` - The timestamp (in milliseconds since UNIX EPOCH) to use.
+    /// We use a compare-and-swap (CAS) technique to attempt to repeatedly build a fresh UUID
+    /// until we succeed. Benchmarks roughly twice as fast as using a Mutex.
     ///
-    /// # Panics
-    ///
-    /// * if the given timestamp is greater than 2^48 - 1.
-    pub(crate) fn build_with_instant(&self, timestamp: u64) -> UUID {
-        assert!(
-            timestamp & MAX_TIMESTAMP_MASK == 0,
-            "Timestamp of UUID must not exceed 48 bits"
-        );
-
-        let new_msb = {
+    pub(crate) fn build_internal(&self) -> UUID {
+        loop {
             let current_msb = self.msb.load(Ordering::SeqCst);
 
-            if timestamp == (current_msb >> 16) {
-                if (current_msb & MAX_COUNT) < MAX_COUNT {
-                    self.msb.fetch_add(1, Ordering::SeqCst);
+            let cas_top_of_loop_time = SystemTime::now()
+                .duration_since(UNIX_EPOCH)
+                .expect("current system time is set to a point in time before UNIX Epoch");
+            let cas_top_of_loop_millis = u64::try_from(cas_top_of_loop_time.as_millis())
+                .expect("current system time is set to a point in time too far in the future");
+
+            let current_timestamp = current_msb >> 16;
+            let new_msb;
+
+            if cas_top_of_loop_millis == current_timestamp {
+                // If the timestamp hasn't changed, attempt to increment the counter.
+                let current_counter = current_msb & MAX_COUNT;
+                if current_counter < MAX_COUNT {
+                    // Prepare new msb with incremented counter.
+                    new_msb = current_msb + 1;
                 } else {
                     // this should never happen in practice because we
                     // do not expect any uEntity to emit more than
                     // 4095 messages/ms
                     // so we simply keep the current counter at MAX_COUNT
+                    new_msb = current_msb
```

Comment:
while it might be true that we do not reach this limit easily, I tend to agree with @AnotherDaniel that we might simply want to wait for the new millenium, eh millisecond, here, instead of issuing the same UUID multiple times.

I guess simply `continue`ing should do the trick here, shouldn't it? The price to pay would be a very slightly delayed sending of messages but at least, we would always have unique UUIDs ...